src/control/lib/control/pool.go

```diff
@@ -1057,89 +1063,212 @@ func processSCMSpaceStats(log debugLogger, filterRank filterRankFn, scmNamespace
+			if !req.filterRank(smdDevice.Rank) {
+				req.log.Debugf("skipping %s, not in ranklist", msgDev)
 				continue
 			}
 
-			if _, exists := rankNVMeFreeSpace[smdDevice.Rank]; !exists {
-				return errors.Errorf("Rank %d without SCM device and at least one %s",
+			if _, exists := req.rankNVMeFreeSpace[smdDevice.Rank]; !exists {
+				return errors.Errorf("rank %d without scm device and at least one %s",
 					smdDevice.Rank, msgDev)
 			}
+			req.rankNVMeFreeSpace[smdDevice.Rank] += smdDevice.UsableBytes
 
-			rankNVMeFreeSpace[smdDevice.Rank] += smdDevice.UsableBytes
+			// TODO DAOS-16160: Adjust size to accommodate for WAL requirements.
 
-			log.Debugf("Added %s as usable: device state=%q, smd-size=%d ctrlr-total-free=%d",
-				msgDev, controller.NvmeState.String(), smdDevice.UsableBytes,
-				rankNVMeFreeSpace[smdDevice.Rank])
+			req.log.Debugf("added %s as usable: device state=%q, smd-size %s (%d), "+
+				"ctrlr-total-free %s (%d)", msgDev, controller.NvmeState.String(),
+				humanize.Bytes(smdDevice.UsableBytes), smdDevice.UsableBytes,
+				humanize.Bytes(req.rankNVMeFreeSpace[smdDevice.Rank]),
+				req.rankNVMeFreeSpace[smdDevice.Rank])
 		}
 	}
 
 	return nil
 }
 
-// Return the maximal SCM and NVMe size of a pool which could be created with all the storage nodes.
-func getMaxPoolSize(ctx context.Context, rpcClient UnaryInvoker, ranks ranklist.RankList) (uint64, uint64, error) {
-	// Verify that the DAOS system is ready before attempting to query storage.
-	if _, err := SystemQuery(ctx, rpcClient, &SystemQueryReq{}); err != nil {
-		return 0, 0, err
+// Derive mem_size and data_size components using percentage of available ramdisk and SSD capacity.
+// The rationale for deriving space usage from ramdisk usage is that this is most likely to be the
+// limiting factor as opposed to SSD usage.
+// - Calculate mem_size as percentage of ramdisk/tmpfs free capacity.
+// - Calculate meta_size as mem_size/mem-ratio.
+// - When calculating data_size in pool…
+//   - Subtract meta_size from free blobstore space if an SSD shares META+DATA roles
+//   - Subtract WAL size from free blobstore space if an SSD shares WAL+DATA roles
+//   - Take percentage of remainder
+func getMaxPoolSizeMdOnSsd(ctx context.Context, rpcClient UnaryInvoker, createReq *PoolCreateReq, scanResp *StorageScanResp) (uint64, uint64, error) {
+	if createReq.MemRatio < 0 {
+		return 0, 0, errors.New("invalid mem-ratio, should be greater than zero")
 	}
+	if createReq.MemRatio > 1 {
+		return 0, 0, errors.New("invalid mem-ratio, should not be greater than one")
+	}
+	ranks := ranklist.RankList(createReq.Ranks)
 
-	resp, err := StorageScan(ctx, rpcClient, &StorageScanReq{Usage: true})
-	if err != nil {
-		return 0, 0, err
+	// Request parameters used to query and calculate space requirements for pool.
+	spaceReq := processSpaceReq{
+		log:               rpcClient,
+		filterRank:        newFilterRankFunc(ranks),
+		rankNVMeFreeSpace: make(rankFreeSpaceMap),
```

Comment:
Seeing how this is used, I'm inclined to pass it separately from the req structure, to make clear it is intended to be modified as a result of the functions it's passed into.

Reply:
I'm going to replace this with seenRanks in order to separate concerns and keep the map local to the NVMe function